File: 06_detectLoop/6.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None

class Solution:
    def detectLoop(self, head):
        slow = head
        fast = head
        while fast and fast.next:
            slow = slow.next
            fast = fast.next.next
            if slow == fast:
                return True
        return False

def build_linked_list_with_loop(values, loop_index):
    if not values:
        return None
    nodes = []
    head = Node(values[0])
    nodes.append(head)
    current = head
    for val in values[1:]:
        new_node = Node(val)
        nodes.append(new_node)
        current.next = new_node
        current = new_node
    if loop_index != -1:
        current.next = nodes[loop_index]
    return head

head = build_linked_list_with_loop([1, 2, 3, 4, 5], 2)
sol = Solution()
print(sol.detectLoop(head))

File: 06_detectLoop/6.c
#include <stdio.h>
#include <stdlib.h>

struct Node {
    int data;
    struct Node* next;
};

int detectLoop(struct Node* head) {
    struct Node* slow = head;
    struct Node* fast = head;
    while (fast && fast->next) {
        slow = slow->next;
        fast = fast->next->next;
        if (slow == fast) return 1;
    }
    return 0;
}

struct Node* build_linked_list_with_loop(int arr[], int n, int loop_index) {
    if (n == 0) return NULL;
    struct Node** nodes = (struct Node**)malloc(n * sizeof(struct Node*));
    struct Node* head = (struct Node*)malloc(sizeof(struct Node));
    head->data = arr[0];
    head->next = NULL;
    nodes[0] = head;
    struct Node* current = head;
    for (int i = 1; i < n; i++) {
        struct Node* new_node = (struct Node*)malloc(sizeof(struct Node));
        new_node->data = arr[i];
        new_node->next = NULL;
        current->next = new_node;
        current = new_node;
        nodes[i] = new_node;
    }
    if (loop_index != -1) {
        current->next = nodes[loop_index];
    }
    free(nodes);
    return head;
}

int main() {
    int arr[] = {1, 2, 3, 4, 5};
    int n = 5;
    int loop_index = 2;
    struct Node* head = build_linked_list_with_loop(arr, n, loop_index);
    printf("%d\n", detectLoop(head));
    return 0;
}